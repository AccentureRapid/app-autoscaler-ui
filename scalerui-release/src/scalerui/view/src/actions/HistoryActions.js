import ActionTypes from '../constants/ActionTypes'
import Sources from '../sources'

let HistoryActions = {

  setHistoryOptions(options) {
    return {
      type: ActionTypes.SET_APP_HISTORY_OPTIONS_SUCCESS,
      options
    }
  },

  getHistory(appId, query) {
    return function (dispatch) {
      dispatch(HistoryActions.getHistoryStart())
      Sources.getAppHistory((err, res) => {
        if (err && !res) {
          //network error
          dispatch(HistoryActions.getHistoryFailed(err.toString()))
        } else {
          if (res.status == 200) {
            dispatch(HistoryActions.getHistorySuccess(res.body, query))
          } else if (res.status == 401) {
            //unauthorized
            dispatch(HistoryActions.getHistoryFailed(res.status + ' ' + res.statusText))
          } else if (res.status == 403) {
            //unbind
            if (!res.body) {
              dispatch(HistoryActions.getHistoryFailed(res.status + ' ' + res.statusText + ' ' + res.text))
            } else {
              dispatch(HistoryActions.getHistoryFailed(res.body.error))
            }
          } else {
            //other error
            dispatch(HistoryActions.getHistoryFailed(res.status + ' ' + res.statusText + ' ' + JSON.stringify(res.body)))
          }
        }
      }, appId, query)
    }
  },

  getHistoryStart() {
    return {
      type: ActionTypes.GET_APP_HISTORY_START
    }
  },

  getHistorySuccess(data, query) {
    return {
      type: ActionTypes.GET_APP_HISTORY_SUCCESS,
      data,
      query
    }
  },

  getHistoryFailed(err) {
    return {
      type: ActionTypes.GET_APP_HISTORY_FAILED,
      err
    }
  },

  resetHistoryView() {
    return {
      type: ActionTypes.RESET_HISTORY_VIEW
    }
  }

}

export default HistoryActions
