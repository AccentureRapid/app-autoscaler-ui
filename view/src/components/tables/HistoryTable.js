import React from 'react'
import { connect } from 'react-redux'
import { IntlProvider, FormattedMessage } from 'react-intl'
import Locale from '../../common/Locale'
import { TableRow, TableData, TableHead, TableBody, TableHeader, Table } from 'carbon-components-react'
import Constants from '../../constants/Constants'
import moment from 'moment-timezone'
import Util from '../../common/Util'

const scaleStatusImages = {
  'succeeded': require('../../images/green_circle.svg'),
  'failed': require('../../images/red_circle.svg'),
  'ignored': require('../../images/orange_circle.svg')
}

class HistoryTable extends React.Component {
  constructor(props) {
    super(props)
  }

  componentWillMount() {
  }

  componentWillUnmount() {
  }

  render() {
    let records = this.props.historyData.get('app_history_data').get('resources').toJS()
    let tableRows = records.map((record, i) => {
      let description = (() => {
        if (record.message) {
          let change = record.new_instances - record.old_instances
          if (change > 0) change = '+' + change
          return change + ' instance(s) because ' + record.message
        } else {
          return record.reason
        }
      })()
      let image = (() => {
        if (record.old_instances > record.new_instances) {
          return <nobr className='scale-in-arrow'>➘ </nobr>
        } else if (record.old_instances < record.new_instances) {
          return <nobr className='scale-out-arrow'>➚ </nobr>
        } else {
          return <nobr className='scale-error-arrow'>➙ </nobr>
        }
      })()
      let timezone = moment.tz.names()[this.props.historyData.get('selected_options').toJS()['timezoneIndex']]
      return (
        <TableRow key={i}>
          <TableData style={{ width: '100%', wordBreak: 'keep-all', whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis' }}>
            <img className='small' src={scaleStatusImages[Constants.ScalingStatus[record.status]]}></img><FormattedMessage id={'history_page_scaling_status_' + Constants.ScalingStatus[record.status]} />
          </TableData>
          <TableData>
            <FormattedMessage id={'history_page_scaling_type_' + Constants.ScalingType[record.scaling_type]} />
          </TableData>
          <TableData>
            {record.old_instances} {image} {record.new_instances}
          </TableData>
          <TableData>
            {moment(record.timestamp / 1000000).tz(timezone).format(Constants.MomentFormateDateTime)}
          </TableData>
          <TableData>
            {description}
          </TableData>
          <TableData>
            {record.error}
          </TableData>
        </TableRow>
      )
    })

    return (
      <IntlProvider locale={Util.getDefaultLocal()} messages={Locale.getLocale(Util.getDefaultLocal())}>
        <div>
          <Table>
            <TableHead>
              <TableRow header>
                <TableHeader>
                  <FormattedMessage id='history_page_scaling_status' />
                </TableHeader>
                <TableHeader>
                  <FormattedMessage id='history_page_scaling_type' />
                </TableHeader>
                <TableHeader>
                  <FormattedMessage id='history_page_header_instance_change' />
                </TableHeader>
                <TableHeader>
                  <FormattedMessage id='history_page_header_time' />
                </TableHeader>
                <TableHeader>
                  <FormattedMessage id='history_page_header_action' />
                </TableHeader>
                <TableHeader>
                  <FormattedMessage id='history_page_header_error' />
                </TableHeader>
              </TableRow>
            </TableHead>
            <TableBody>
              {tableRows}
            </TableBody>
          </Table>
        </div>
      </IntlProvider>
    )
  }
}

export default connect(
  state => ({
    historyData: state.historyData
  })
)(HistoryTable)