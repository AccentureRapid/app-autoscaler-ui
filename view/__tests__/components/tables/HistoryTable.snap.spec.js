import HistoryTable from '../../../src/components/tables/HistoryTable'
import renderer from 'react-test-renderer'
import HistoryActions from '../../../src/actions/HistoryActions'
import store from '../../../src/store'
import fs from 'fs'
import React from 'react'
import TestConstants from '../../TestConstants'
import moment from 'moment-timezone'

const historyFile = '__tests__/assets/history/history.json'
const historyData = JSON.parse(fs.readFileSync(historyFile))
const selectedOptions = {
  range: {
    typeIndex: 0,
    from: 0,
    to: 0
  },
  timezoneIndex: moment.tz.names().indexOf(TestConstants.timezone),
  scalingStatusIndex: 0,
  scalingTypeIndex: 0,
  scalingActionIndex: 0,
  perPage: 10
}
store.dispatch(HistoryActions.setHistoryOptions(selectedOptions))
store.dispatch(HistoryActions.getHistorySuccess(historyData))

describe('HistoryTable', () => {
  it.only('renders HistoryTable component correctly', () => {
    const wrapper = renderer.create(
      <HistoryTable store={store} />
    )
    const tree = wrapper.toJSON()
    expect(tree).toMatchSnapshot()
    wrapper.unmount()
  })
})
