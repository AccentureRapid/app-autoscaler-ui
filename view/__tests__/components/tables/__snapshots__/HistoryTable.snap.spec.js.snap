// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryTable renders HistoryTable component correctly 1`] = `
<div>
  <div
    className="bx--responsive-table-container"
  >
    <table
      className="bx--responsive-table"
      style={
        Object {
          "borderCollapse": "collapse",
          "borderSpacing": 0,
        }
      }
    >
      <thead
        className="bx--table-head"
      >
        <tr
          className="bx--table-row"
        >
          <th
            className="bx--table-header"
          >
            <span>
              Scaling Status
            </span>
            
          </th>
          <th
            className="bx--table-header"
          >
            <span>
              Scaling Type
            </span>
            
          </th>
          <th
            className="bx--table-header"
          >
            <span>
              Instance Change
            </span>
            
          </th>
          <th
            className="bx--table-header"
          >
            <span>
              Time
            </span>
            
          </th>
          <th
            className="bx--table-header"
          >
            <span>
              Action
            </span>
            
          </th>
          <th
            className="bx--table-header"
          >
            <span>
              Error
            </span>
            
          </th>
        </tr>
      </thead>
      <tbody
        className="bx--table-body"
      >
        <tr
          className="bx--table-row bx--parent-row bx--parent-row--even"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Dynamic
            </span>
          </td>
          <td
            className=""
          >
            2
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            1
          </td>
          <td
            className=""
          >
            2018-05-15 02:42:55
          </td>
          <td
            className=""
          >
            -1 instance(s) because memoryused &lt; 30MB for 600 seconds
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
        <tr
          className="bx--table-row bx--parent-row"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Dynamic
            </span>
          </td>
          <td
            className=""
          >
            3
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            2
          </td>
          <td
            className=""
          >
            2018-05-15 02:37:35
          </td>
          <td
            className=""
          >
            -1 instance(s) because throughput &lt; 5rps for 60 seconds
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
        <tr
          className="bx--table-row bx--parent-row bx--parent-row--even"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Dynamic
            </span>
          </td>
          <td
            className=""
          >
            4
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            3
          </td>
          <td
            className=""
          >
            2018-05-15 02:32:15
          </td>
          <td
            className=""
          >
            -1 instance(s) because throughput &lt; 5rps for 60 seconds
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
        <tr
          className="bx--table-row bx--parent-row"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Dynamic
            </span>
          </td>
          <td
            className=""
          >
            5
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            4
          </td>
          <td
            className=""
          >
            2018-05-15 02:26:55
          </td>
          <td
            className=""
          >
            -1 instance(s) because memoryutil &lt; 30% for 60 seconds
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
        <tr
          className="bx--table-row bx--parent-row bx--parent-row--even"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Dynamic
            </span>
          </td>
          <td
            className=""
          >
            6
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            5
          </td>
          <td
            className=""
          >
            2018-05-15 02:21:34
          </td>
          <td
            className=""
          >
            -1 instance(s) because memoryutil &lt; 30% for 60 seconds
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
        <tr
          className="bx--table-row bx--parent-row"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Dynamic
            </span>
          </td>
          <td
            className=""
          >
            7
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            6
          </td>
          <td
            className=""
          >
            2018-05-15 02:16:15
          </td>
          <td
            className=""
          >
            -1 instance(s) because throughput &lt; 5rps for 60 seconds
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
        <tr
          className="bx--table-row bx--parent-row bx--parent-row--even"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Dynamic
            </span>
          </td>
          <td
            className=""
          >
            8
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            7
          </td>
          <td
            className=""
          >
            2018-05-15 02:10:55
          </td>
          <td
            className=""
          >
            -1 instance(s) because throughput &lt; 5rps for 60 seconds
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
        <tr
          className="bx--table-row bx--parent-row"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Dynamic
            </span>
          </td>
          <td
            className=""
          >
            9
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            8
          </td>
          <td
            className=""
          >
            2018-05-15 02:05:35
          </td>
          <td
            className=""
          >
            -1 instance(s) because memoryutil &lt; 30% for 60 seconds
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
        <tr
          className="bx--table-row bx--parent-row bx--parent-row--even"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Dynamic
            </span>
          </td>
          <td
            className=""
          >
            10
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            9
          </td>
          <td
            className=""
          >
            2018-05-15 02:00:15
          </td>
          <td
            className=""
          >
            -1 instance(s) because throughput &lt; 5rps for 60 seconds
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
        <tr
          className="bx--table-row bx--parent-row"
        >
          <td
            className=""
            style={
              Object {
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
                "wordBreak": "keep-all",
              }
            }
          >
            <img
              className="small"
              src={
                Object {
                  "default": Object {
                    "mock": "mock",
                  },
                }
              }
            />
            <span>
              Succeeded
            </span>
          </td>
          <td
            className=""
          >
            <span>
              Schedule
            </span>
          </td>
          <td
            className=""
          >
            18
             
            <nobr
              className="scale-in-arrow"
            >
              ➘ 
            </nobr>
             
            10
          </td>
          <td
            className=""
          >
            2018-05-15 02:00:03
          </td>
          <td
            className=""
          >
            -8 instance(s) because limited by max instances 5
          </td>
          <td
            className=""
          >
            
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
